metadata-io/src/main/java/com/linkedin/metadata/service/DataProductService.java

```diff
@@ -0,0 +1,387 @@
+package com.linkedin.metadata.service;
+
+import com.datahub.authentication.Authentication;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.linkedin.common.AuditStamp;
+import com.linkedin.common.EntityRelationships;
+import com.linkedin.common.UrnArray;
+import com.linkedin.common.urn.Urn;
+import com.linkedin.common.urn.UrnUtils;
+import com.linkedin.data.DataMap;
+import com.linkedin.data.template.SetMode;
+import com.linkedin.dataproduct.DataProductAssociation;
+import com.linkedin.dataproduct.DataProductAssociationArray;
+import com.linkedin.dataproduct.DataProductKey;
+import com.linkedin.dataproduct.DataProductProperties;
+import com.linkedin.domain.Domains;
+import com.linkedin.entity.EntityResponse;
+import com.linkedin.entity.client.EntityClient;
+import com.linkedin.metadata.Constants;
+import com.linkedin.metadata.entity.AspectUtils;
+import com.linkedin.metadata.graph.GraphClient;
+import com.linkedin.metadata.query.filter.RelationshipDirection;
+import com.linkedin.metadata.utils.EntityKeyUtils;
+import com.linkedin.r2.RemoteInvocationException;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to permit easy CRUD operations on a DataProduct
+ *
+ * Note that no Authorization is performed within the service. The expectation
+ * is that the caller has already verified the permissions of the active Actor.
+ *
+ */
+@Slf4j
+public class DataProductService {
+  private final EntityClient _entityClient;
+  private final GraphClient _graphClient;
+
+  public DataProductService(@Nonnull EntityClient entityClient, @Nonnull GraphClient graphClient) {
+    _entityClient = entityClient;
+    _graphClient = graphClient;
+  }
+
+  /**
+   * Creates a new Data Product.
+   *
+   * Note that this method does not do authorization validation.
+   * It is assumed that users of this class have already authorized the operation.
+   *
+   * @param name optional name of the DataProduct
+   * @param description optional description of the DataProduct
+   *
+   * @return the urn of the newly created DataProduct
+   */
+  public Urn createDataProduct(
+      @Nullable String name,
+      @Nullable String description,
+      @Nonnull Authentication authentication) {
+
+    // 1. Generate a unique id for the new DataProduct.
```

Comment:
Should we allow a user to provide a unique id for their product? Looks likke shirshanka's YAML format would allow that

Reply:
ahh good point - I could definitely see it, but maybe come back to this as a followup since i don't think that's essential in V1

Reply:
okay - so long as we keep track !